packages/paste-core/components/chat-log/src/useChatLog.ts

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import {uid as buildID} from '@twilio-paste/uid-library';
+
+type ChatDictionary = Record<string, ChatData>;
+type PushChat = (chat: ChatData) => void;
+type PopChat = (id: string) => void;
+type Chats = React.ReactElement[];
+
+export type ChatData = {id?: string; content: React.ReactElement};
+export type UseChatLog = (...initialChats: ChatData[]) => {
+  chats: Chats;
+  chatIds: string[];
+  push: PushChat;
+  pop: PopChat;
+};
+
+const spreadChatIntoDictionary = (prev: ChatDictionary, next: ChatData): ChatDictionary => ({
+  ...prev,
+  [next.id || buildID(next.content)]: next,
+});
+
+export const useChatLog: UseChatLog = (...initialChats) => {
+  const parsedInitialChats = React.useMemo(
+    () => initialChats.reduce((prev, next) => spreadChatIntoDictionary(prev, next), {}),
+    [initialChats]
+  );
+
+  const [chatDictionary, setChatDictionary] = React.useState<ChatDictionary>(parsedInitialChats);
+
+  const push: PushChat = React.useCallback(
+    (next) => {
+      setChatDictionary((prev) => spreadChatIntoDictionary(prev, next));
+    },
+    [setChatDictionary]
+  );
+
+  const pop: PopChat = React.useCallback(
+    (id) => {
+      setChatDictionary((prev) => {
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const {[id]: _, ...restPrev} = prev;
+        return restPrev;
+      });
+    },
+    [setChatDictionary]
+  );
+
+  const parsedChatData = React.useMemo(
```

Comment:
Question: Could you explain how this function works? Sorry, I'm having some difficulty understanding.